File: src/main/java/com/leijendary/spring/microservicetemplate/model/SnowflakeIdModel.java
package com.leijendary.spring.microservicetemplate.model;

import lombok.Data;
import lombok.EqualsAndHashCode;
import org.hibernate.annotations.GenericGenerator;

import javax.persistence.GeneratedValue;
import javax.persistence.Id;
import javax.persistence.MappedSuperclass;

@Data
@EqualsAndHashCode(callSuper = true)
@MappedSuperclass
public abstract class SnowflakeIdModel extends AbstractModel {

    protected static final String GENERATOR_SNOWFLAKE = "snowflake";

    @Id
    @GenericGenerator(
            name = "snowflake",
            strategy = "com.leijendary.spring.microservicetemplate.model.generator.SnowflakeSequenceGenerator")
    @GeneratedValue(generator = GENERATOR_SNOWFLAKE)
    private long id;
}

File: src/main/kotlin/com/leijendary/spring/template/core/model/SnowflakeIdModel.kt
package com.leijendary.spring.template.core.model

import com.leijendary.spring.template.core.generator.SnowflakeIdGenerator.Companion.STRATEGY
import jakarta.persistence.GeneratedValue
import jakarta.persistence.Id
import jakarta.persistence.MappedSuperclass
import org.hibernate.annotations.GenericGenerator

@MappedSuperclass
abstract class SnowflakeIdModel : AppModel() {
    companion object {
        protected const val GENERATOR_SNOWFLAKE = "snowflake"
    }

    @Id
    @GeneratedValue(generator = GENERATOR_SNOWFLAKE)
    @GenericGenerator(name = GENERATOR_SNOWFLAKE, strategy = STRATEGY)
    var id: Long = 0
}